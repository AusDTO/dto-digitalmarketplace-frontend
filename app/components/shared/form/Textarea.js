import React from 'react'
import PropTypes from 'prop-types'
import { Control, controls } from 'react-redux-form'

import StatefulError from './StatefulError'
import TextareaComponent from './Textarea/Textarea'

import { limitWords } from '../../validators'

const Textarea = props => {
  const { name, id, label, model, description, showMessagesDuringFocus = false, controlProps = {}, mapProps } = props

  let { validators, messages } = props

  if (controlProps.limit) {
    validators = { ...validators, limitWords: limitWords(controlProps.limit) }
    messages = {
      limitWords: `${label} has exceeded the word limit.`,
      ...messages
    }
  }
  return (
    <div className="field">
      <label className="question-heading" htmlFor={id}>
        {label}
      </label>
      {description &&
        <p className="hint" id={`${id}-hint`}>
          {description}
        </p>}
      {messages &&
        <StatefulError model={model} messages={messages} id={id} showMessagesDuringFocus={showMessagesDuringFocus} />}
      <Control
        id={id}
        model={model}
        controlProps={{
          name,
          id,
          describedby: `${id}-hint`,
          hint: description,
          ...controlProps
        }}
        validators={validators}
        component={TextareaComponent}
        mapProps={{
          className: ({ fieldValue }) =>
            !fieldValue.valid && fieldValue.touched
              ? 'uikit-text-input--invalid uikit-text-input uikit-text-input--block uikit-text-input--textarea'
              : 'uikit-text-input uikit-text-input--block uikit-text-input--textarea',
          value: prps => prps.viewValue,
          ...mapProps,
          ...controls.default
        }}
      />
    </div>
  )
}

Textarea.defaultProps = {
  mapProps: null,
  validators: null,
  messages: null,
  description: '',
  controlProps: null
}

Textarea.propTypes = {
  name: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  model: PropTypes.oneOfType([PropTypes.func, PropTypes.string]).isRequired,

  validators: PropTypes.object,
  messages: PropTypes.object,
  description: PropTypes.string,
  controlProps: PropTypes.object,
  mapProps: PropTypes.oneOfType([PropTypes.func, PropTypes.object])
}

export default Textarea
