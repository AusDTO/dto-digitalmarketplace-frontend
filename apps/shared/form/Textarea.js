import React from 'react'
import PropTypes from 'prop-types'
import { Control, controls } from 'react-redux-form'
import { limitWords } from 'shared/validators'

import StatefulError from './StatefulError'
import TextareaComponent from './Textarea/Textarea'

const Textarea = props => {
  const { name, id, label, model, description, showMessagesDuringFocus = true, controlProps = {}, mapProps } = props

  let { validators, messages } = props

  if (controlProps.limit) {
    validators = { ...validators, limitWords: limitWords(controlProps.limit) }
    if (!messages || !messages.limitWords) {
      messages = {
        limitWords: ` ${label} has exceeded the word limit.`,
        ...messages
      }
    }
  }
  return (
    <div className="field">
      <label id={`${id}-label`} className="question-heading au-text-input__label" htmlFor={id}>
        {label}
      </label>
      {description &&
        <div className="au-text-input__hint" id={`${id}-hint`}>
          {description}
        </div>}
      {messages &&
        <StatefulError model={model} messages={messages} id={id} showMessagesDuringFocus={showMessagesDuringFocus} />}
      <Control
        id={id}
        model={model}
        controlProps={{
          name,
          id,
          describedby: description ? `${id}-hint` : `${id}-label`,
          hint: description,
          ...controlProps
        }}
        validators={validators}
        component={TextareaComponent}
        mapProps={{
          className: ({ fieldValue }) =>
            !fieldValue.valid && fieldValue.touched
              ? 'au-text-input--invalid au-text-input au-text-input--block au-text-input--textarea'
              : 'au-text-input au-text-input--block au-text-input--textarea',
          value: prps => prps.viewValue,
          ...mapProps,
          ...controls.default
        }}
      />
    </div>
  )
}

Textarea.defaultProps = {
  mapProps: null,
  validators: null,
  messages: null,
  description: '',
  controlProps: null
}

Textarea.propTypes = {
  name: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  model: PropTypes.oneOfType([PropTypes.func, PropTypes.string]).isRequired,

  validators: PropTypes.object,
  messages: PropTypes.object,
  description: PropTypes.string,
  controlProps: PropTypes.object,
  mapProps: PropTypes.oneOfType([PropTypes.func, PropTypes.object])
}

export default Textarea
