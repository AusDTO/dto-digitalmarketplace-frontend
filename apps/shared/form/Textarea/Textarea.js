/* eslint-disable */
import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

class Textarea extends Component {
  static defaultProps = {
    onChange: () => {},
    onBlur: () => {},
    onFocus: () => {}
  }

  constructor(props) {
    super(props)
    const { value = '', limit = 0 } = props
    const words = this.countWords(value)

    this.state = {
      wordsLeft: limit - words
    }
  }

  countWords(s: string) {
    const count = (s && s.match(/\S+/g)) || []
    return count.length
  }

  onChange(e) {
    const content = e.target.value
    const words = this.countWords(content)
    const { limit, onChange } = this.props
    this.setState({
      wordsLeft: limit - words
    })
    onChange(content)
  }

  limitText(counter: number, wordsLeft: number) {
    let words = 'words'
    let affix = 'left'

    if (counter === 1 || counter === -1) {
      words = 'word'
    }

    if (wordsLeft < 0) {
      affix = 'too many'
    }

    return `${counter} ${words} ${affix}`
  }

  render() {
    let { value, limit, name, id, onBlur, onFocus, className = '', describedby, hint } = this.props
    let { wordsLeft } = this.state

    let counter = wordsLeft
    if (counter < 0) {
      counter *= -1
    }

    if (limit) {
      className = classNames(className, {
        'under-50': limit <= 50,
        'under-100': limit <= 100 && limit > 50,
        'under-200': limit <= 200 && limit > 100,
        'over-200': limit > 200
      })
    }

    return (
      <div>
        <textarea
          className={className}
          name={name}
          id={id}
          value={value}
          onBlur={onBlur}
          onFocus={onFocus}
          aria-describedby={describedby}
          onChange={this.onChange.bind(this)}
        />
        {limit
          ? <span className="word-count-counter" aria-live="polite">
              {this.limitText(counter, wordsLeft)}
            </span>
          : ''}
      </div>
    )
  }
}

Textarea.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string.isRequired,
  limit: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  value: PropTypes.string
}

export default Textarea
