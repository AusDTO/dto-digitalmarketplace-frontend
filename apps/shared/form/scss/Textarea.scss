.wordCount {
  display: block;
  text-align: right;
  color: #717171;
}