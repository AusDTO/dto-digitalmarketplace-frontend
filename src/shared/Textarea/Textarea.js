import React, { Component, PropTypes } from 'react'

class Textarea extends Component {

  static defaultProps = {
    onChange: () => {},
    onBlur: () => {},
    onFocus: () => {}
  }

  constructor(props) {
    super(props)
    const { value = '', limit = 0 } = props
    const words = this.countWords(value)

    this.state = {
      wordsLeft: limit - words,
    }
  }

  countWords(s: string) {
    const count =  s.match(/\S+/g) || [];
    return count.length;
  }

  onChange(e) {
    const content = e.target.value
    const words = this.countWords(content)
    const { limit, onChange } = this.props
    this.setState({
      wordsLeft: limit - words,
    })
    onChange(content);
  }

  limitText(counter: number, wordsLeft: number) {
    let words = 'words'
    let affix = 'remaining'

    if (counter === 1 || counter === -1) {
      words = 'word'
    }

    if (wordsLeft < 0) {
      affix = 'too many'
    }

    return `${counter} ${words} ${affix}`
  }

  render() {
    const { value, limit, name, id, onBlur, onFocus, className = '' } = this.props
    let { wordsLeft } = this.state

    let counter = wordsLeft
    if (counter < 0) {
      counter *= -1
    }

    return (
      <div>
        <textarea
          ref="textarea"
          className={className}
          name={name}
          id={id}
          defaultValue={value}
          onBlur={onBlur}
          onFocus={onFocus}
          onChange={this.onChange.bind(this)}
        ></textarea>
        {limit ? (
          <span className="limit">{this.limitText(counter, wordsLeft)}</span>
        ) : ''}
      </div>
    )
  }
}

Textarea.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string.isRequired,
  limit: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  value: PropTypes.string
}

export default Textarea
