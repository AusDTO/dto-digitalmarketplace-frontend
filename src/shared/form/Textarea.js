import React from 'react';
import { Control, controls } from 'react-redux-form';

import StatefulError from './StatefulError';
import TextareaComponent from '../Textarea';

const Textarea = ({ name, id, label, model, validators, messages, description, controlProps, mapProps }) => (
  <div className="field">
    <label className="question-heading" htmlFor={id}>{label}</label>
    {description && (
      <p className="hint" id={`${id}-hint`}>{description}</p>
    )}
    {messages && <StatefulError model={model} messages={messages} id={id} />}
    <Control
      model={model}
      controlProps={{ name, id, describedby: `${id}-hint`, hint: description, ...controlProps}}
      validators={validators}
      component={TextareaComponent}

      mapProps={{
        className: ({ fieldValue }) => !fieldValue.valid && fieldValue.touched ? 'invalid' : '',
        value: (props) => props.viewValue,
        ...mapProps,
        ...controls.default,
      }}
    />
  </div>
);

Textarea.defaultProps = {
  mapProps: {}
}

Textarea.propTypes = {
  name: React.PropTypes.string.isRequired,
  id: React.PropTypes.string.isRequired,
  label: React.PropTypes.string.isRequired,
  model: React.PropTypes.oneOfType([
    React.PropTypes.func,
    React.PropTypes.string,
  ]).isRequired,

  validators: React.PropTypes.object,
  messages: React.PropTypes.object,
  description: React.PropTypes.string,
  controlProps: React.PropTypes.object,
  mapProps: React.PropTypes.oneOfType([
    React.PropTypes.func,
    React.PropTypes.object,
  ]),
};


export default Textarea;