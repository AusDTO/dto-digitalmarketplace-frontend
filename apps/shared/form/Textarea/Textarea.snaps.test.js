import React from 'react'
import renderer from 'react-test-renderer'

import Textarea from './Textarea'

test('Textarea with no words, no limit', () => {
  const component = renderer.create(<Textarea name="test" />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Textarea with limit 100, no words', () => {
  const component = renderer.create(<Textarea name="test" limit={100} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Textarea with limit 100, 10 words', () => {
  const component = renderer.create(<Textarea name="test" limit={100} value="1 2 3 4 5 6 7 8 9 10" />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Textarea with limit 5, 10 words', () => {
  const component = renderer.create(<Textarea name="test" limit={5} value="1 2 3 4 5 6 7 8 9 10" />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
