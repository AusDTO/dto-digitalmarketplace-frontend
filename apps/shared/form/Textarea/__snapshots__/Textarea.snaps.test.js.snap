// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea with limit 5, 10 words 1`] = `
<div>
  <textarea
    aria-describedby={undefined}
    className="under-50"
    id={undefined}
    name="test"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value="1 2 3 4 5 6 7 8 9 10"
  />
  <span
    aria-live="polite"
    className="word-count-counter"
  >
    5 words too many
  </span>
</div>
`;

exports[`Textarea with limit 100, 10 words 1`] = `
<div>
  <textarea
    aria-describedby={undefined}
    className="under-100"
    id={undefined}
    name="test"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value="1 2 3 4 5 6 7 8 9 10"
  />
  <span
    aria-live="polite"
    className="word-count-counter"
  >
    90 words left
  </span>
</div>
`;

exports[`Textarea with limit 100, no words 1`] = `
<div>
  <textarea
    aria-describedby={undefined}
    className="under-100"
    id={undefined}
    name="test"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value={undefined}
  />
  <span
    aria-live="polite"
    className="word-count-counter"
  >
    100 words left
  </span>
</div>
`;

exports[`Textarea with no words, no limit 1`] = `
<div>
  <textarea
    aria-describedby={undefined}
    className=""
    id={undefined}
    name="test"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value={undefined}
  />
  
</div>
`;
